wave_containers/include/wave/containers/measurement_container_base.hpp

```diff
@@ -0,0 +1,195 @@
+     */
+    template <typename InputIt>
+    void insert(InputIt first, InputIt last);
+
+    /** Insert a Measurement constructed from the arguments if a measurement for
+     * the same time and sensor does not already exist.
+     *
+     * @return a pair p. If and only if insertion occurred, p.second is true and
+     * p.first points to the element inserted.
+     */
+    template <typename... Args>
+    std::pair<iterator, bool> emplace(Args &&... args);
+
+    /** Delete the element with the matching key, if one exists.
+     *
+     * @param args
+     * @parblock
+     * arguments forming a unique key for the container.
+     *
+     * See the documentation for the derived container class for a definition of
+     * the unique key.
+     *
+     * For example, for MeasurementContainer, the key consists of time and
+     * sensor_id. For LandmarkMeasurementContainer, the key consists of time,
+     * sensor_id, and landmark_id.
+     * @endparblock
+     *
+     * @return the number of elements deleted (0 or 1).
+     */
+    template <typename... Args>
+    size_type erase(Args &&... args);
+
+    /** Delete the element at `position`
+     *
+     * @param position a valid dereferenceable iterator of this container
+     * @return An iterator pointing to the element following the deleted one, or
+     * `end()` if it was the last.
+     */
+    iterator erase(iterator position) noexcept;
+
+
+    /** Delete the elements in the range [first, last)
+     *
+     * @param first, last a valid range of this container
+     * @return `last`
+     */
+    iterator erase(iterator first, iterator last) noexcept;
+
+
+    /** Delete all elements */
+    void clear() noexcept;
+
+    // Retrieval
+
+    /** Get all measurements from the given sensor
+     *
+     * @return a pair of iterators representing the start and end of the range.
+     * If the range is empty, both iterators will be equal.
+     *
+     * @note because these iterators use the underlying ordered index of
+     * sensor_ids, they are not the same type as those from `begin()`,
+     * `getTimeWindow()`, etc.
+     */
+    std::pair<sensor_iterator, sensor_iterator> getAllFromSensor(
+      const SensorIdType &s) const noexcept;
+
+    /** Get all measurements between the given times.
+     *
+     * @param start, end an inclusive range of times, with start <= end
+     *
+     * @return a pair of iterators representing the start and end of the range.
+     * If the range is empty, both iterators will be equal.
+     */
+    std::pair<iterator, iterator> getTimeWindow(const TimeType &start,
+                                                const TimeType &end) const
+      noexcept;
+
+
+    // Iterators
+
+    iterator begin() noexcept;
+    iterator end() noexcept;
+    const_iterator begin() const noexcept;
+    const_iterator end() const noexcept;
+    const_iterator cbegin() const noexcept;
+    const_iterator cend() const noexcept;
+
+ protected:
```

Comment:
Hmm maybe. I sorted them by types and data members but I actually have a typedef down here.. good catch